### PR TITLE
Bootstrap margin fee exchange configuration

### PR DESCRIPTION
**Pull request by dazito**
This pull request replaces https://github.com/agonyforge/arbitrader/pull/320

### CODE REVIEW

**Comment by dazito** [comment]
@AllirionX please have a look at my changes in the `PaperTradeService`. I had a merge conflict there. Direct link to the diff: https://github.com/agonyforge/arbitrader/pull/351/files#diff-6b843429246f24a5598b1910f6350f8e1cd50476dec55d1f4ddcec2b695cc0a4

File: src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java
Comment (AllirionX) [inline src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: A few things here:
- about the logs, the paper exchange is just an exchange, it does not need to know if it's a long or short exchange, only if it's a BID or ASK order. Also the logs say `exchange:{}|exchange:{}` instead of `exchange:{}|margin:{}` I also try to start the paper trading logs by `{} paper exchange:` so we can identify logs coming from the paper trading internal logic easily
- the way the fees are calculated, the margin fees will be used for all orders on the short exchange, even if margin is not needed (if we have sufficient funds to sell, or buy I think we should only pay the trade fees)

I don't now if it's how it's supposed to work: if we set the leverage when buying on the short exchange, are we supposed to pay margin fees? I would think so, but then the setLeverage(2) in the TradingService.executeOrderPair method should not be set for the short exchange BUY orders - as we don't need the leverage.

A better way to do this would be to check if a margin trade is required or not 
Create 
```
boolean isMarginRequired (LimitOrder order) {
        BigDecimal counterDelta = getCounterDelta(order);
        BigDecimal baseDelta = getBaseDelta(order);
 
        //Margin required because we don't have enough cash
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().counter).add(counterDelta).compareTo(BigDecimal.ZERO) <0)
            return true;

        //Margin required because we dont have enough crypto 
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().base).add(baseDelta).compareTo(BigDecimal.ZERO) <0)
            return true;

        //Margin required because the leverage was set
        if(order.getLeverage() != null)
            return true;

        return false;
    }
```
Change checkBalance (LimitOrder order) to
```
private void checkBalance(LimitOrder order) {
        if(isMarginRequired(order) && !exchangeService.getExchangeMetadata(exchange).getMargin())
            throw new FundsExceededException();
    }
```
And use `isMarginRequired` to see if we should use the margin fees.
```
fee = isMarginRequired(order) && exchangeFee.getMarginFee().isPresent() ? exchangeFee.getMarginFee().get() : exchangeFee.getTradeFee();
LOGGER.info("{} paper exchange: margin enabled:{}|margin required:{}| order using {} fee", exchangeName, exchangeConfiguration.getMargin(), isMarginRequired(order), fee);
```

**Review by AllirionX — CHANGES_REQUESTED** [review]

File: src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java
Comment (dazito) [reply src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: Thanks for the nice feedback :+1: 

I think we always pay margin fees even if we have enough balance for the borrowed coins (short), @scionaltera can you please confirm?

As far as I know setting value`2` to `order.getLeverage()` is only needed for Liquid exchange. @scionaltera can you confirm this as well?

Comment (scionaltera) [reply src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: The exchange should decide whether it's a margin or non-margin trade by looking at `LimitOrder.getLeverage()`. If it's not null, then it's a margin trade (regardless of long or short) and margin fees should be charged. If we really wanted to do it right for margin orders we wouldn't take the full amount out of the balance either. On a 2:1 trade (`getLeverage() == 2`) we'd only take half of the value away from the balance. It's not that important for Arbitrader at this time because we only ever set it to 2 and we never trade above our actual balance.

Comment (AllirionX) [reply src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: In that case, we can have a method `useMargin` to determine if we charge the margin fees or not.

```
private void useMargin(LimitOrder order) {
    return order.getLeverage() != null && exchangeService.getExchangeMetadata(exchange).getMargin() && exchangeFee.getMarginFee().isPresent() ;
}
```
The fees could be:
```
fee = useMargin(order) ? exchangeFee.getMarginFee().get() : exchangeFee.getTradeFee();
LOGGER.info("{} paper exchange: margin enabled:{}|margin required:{}| order using {} fee", exchangeName, exchangeConfiguration.getMargin(), useMargin(order), fee);
```

and change the checkBalance method accordingly

```
private void checkBalance(LimitOrder order) {
    if(!useMargin(order)) {
        BigDecimal counterDelta = getCounterDelta(order);
        BigDecimal baseDelta = getBaseDelta(order);
 
        //Margin required because we don't have enough cash
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().counter).add(counterDelta).compareTo(BigDecimal.ZERO) <0)
            throw new FundsExceededException();

        //Margin required because we dont have enough crypto 
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().base).add(baseDelta).compareTo(BigDecimal.ZERO) <0)
            throw new FundsExceededException();
    } else {
         //TODO after implementing leverage in the paper exchange, we should check here if we have sufficient funds to cover the leverage
    }
}
```

Comment (scionaltera) [reply src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: One other wrinkle I just thought of - the PaperExchange should probably check whether it's configured with `margin: true` or not and reject margin orders if it is not supposed to be a margin-capable exchange. It would be a bug if we tried to submit a margin order to a non-margin exchange, but stranger things have happened.

File: application.example.yaml
Comment (scionaltera) [inline application.example.yaml:173]: This change to the comment is not accurate. The trade fee is applied to every trade. Long or short, margin or not.

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (scionaltera) [inline src/main/java/com/r307/arbitrader/service/TradingService.java:185]: This should be `shortExchangeFee.getTradeFee().add(shortExchangeFee.getMarginFee().get())`. The trade fee and margin fees need to get added together.

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (scionaltera) [inline src/main/java/com/r307/arbitrader/service/TradingService.java:486]: Again if it's a short trade the trade fee and margin fee need to be added together.

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (scionaltera) [inline src/main/java/com/r307/arbitrader/service/TradingService.java:516]: If it's a margin trade the trade fee and margin fee need to be added together.

File: src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java
Comment (scionaltera) [inline src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:299]: The fee on a margin trade needs to be the trade fee + the margin fee.

**Review by scionaltera — CHANGES_REQUESTED** [review]
Lots of comments but only one actual problem. The way this code is written we might set `tradeFee: 0.0026` and `marginFee: 0.0002` based on the examples in the example configuration and the bot would return 0.0002 as the fee for the _entire short trade_ which is incorrect. It should be 0.0028. I have pointed out all the places I saw where this needs to be fixed.

File: src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java
Comment (AllirionX) [reply src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: It does check it with exchangeService.getExchangeMetadata(exchange).getMargin(). Currently, it rejects orders that would make any balance negative unless margin is enabled, via the check balance method

File: application.example.yaml
Comment (dazito) [reply application.example.yaml:181]: Aha, I didn't know that. I will fix it.

File: src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java
Comment (dazito) [reply src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:301]: Thanks @AllirionX for your great feedback!
@scionaltera yeah, I have implemented a check condition for that use case.

**Comment by dazito** [comment]
I think we should not merge this branch. According to: https://github.com/knowm/XChange/pull/3945
Kraken has some memory issues with this version

**Review by scionaltera — DISMISSED** [review]

File: src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java
Comment (AllirionX) [inline src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java:231]: If this checks for funds, for margin enabled, etc. should we rename both checkBalance to verifyOrder ?

File: src/test/java/com/r307/arbitrader/service/paper/PaperTradeServiceTest.java
Comment (AllirionX) [inline src/test/java/com/r307/arbitrader/service/paper/PaperTradeServiceTest.java:114]: This test throws an exception because we did not set the order leverage. 
We should have a test the happy scenario (with `order.getLeverage() != null` ) that does not throw an exception

Comment (dazito) [reply src/test/java/com/r307/arbitrader/service/paper/PaperTradeServiceTest.java:114]: Thanks, you were right. We were missing copying the leverage value and I also created an unit test to assert on that

**Review by scionaltera — APPROVED** [review]
Once master is merged and conflicts resolved I think this should be OK to merge.

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (AllirionX) [inline src/main/java/com/r307/arbitrader/service/TradingService.java:154]: The variable names confused me:
ExchangeFee has two fees: tradeFee and marginFee
But here:
tradeFee is the long exchange tradeFee
marginFee is the sum of tradeFee and marginFee for the longExchange

To avoid confusion we could use different variable names:
longTradeFee (= tradeFee)
shortTotalFee (= tradeFee + marginFee)

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (AllirionX) [inline src/main/java/com/r307/arbitrader/service/TradingService.java:187]: See other comment. 
tradeFee is the long exchange trade fee, and marginFee is the sum of the short exchange tradeFee and marginFee. 
They should be logged as such:
LOGGER.debug("Long trade fee percent: {}", longTradeFee);
LOGGER.debug("Short trade and margin fee percent: {} + {} = {}", shortTradeFee, shortMarginFee, shortTotalFee);

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (AllirionX) [inline src/main/java/com/r307/arbitrader/service/TradingService.java:342]: See other comment, the same naming convention should apply here as well.

File: src/main/java/com/r307/arbitrader/service/TradingService.java
Comment (scionaltera) [reply src/main/java/com/r307/arbitrader/service/TradingService.java:154]: Agreed. Once you take the `marginFee` and add the `tradeFee` to it, it is specific to the short trade and it is the total of both the fees that apply to that trade.

**Review by AllirionX — APPROVED** [review]

**Review by scionaltera — DISMISSED** [review]

**Review by scionaltera — APPROVED** [review]

**Pull request merged**
